print('1. Обработчики исключений')

# Обработчик исключения используется для отлавливания ошибок заранее. Выглядит это следующим образом:

try: # Мы говорим что хотим обработать какой то кусочек кода
    num = 1
    print(1 + 'hello')
except TypeError: # При припотенциальной ошибке мы укажем тип ошибки, и что будет происходить после её появления
    print("Ошибка типа!") # Можно на это смотроеть как на обычную проверку на условие


try:
    x = int(input('Введите число: '))
    r = 5/x
    print(r)
except ZeroDivisionError:
    print('Вы пытаетесь делить на ноль!')
except ValueError:
    print('Ошибка типа!')






print('2.Finally')

try:
    x = int(input('Введите число: '))
    r = 5/x
    print(r)
except ZeroDivisionError:
    print('Вы пытаетесь делить на ноль!')
finally:
    print('Выполнение кода окончено.')






print('3.Try else')

try:
    x = int(input('Введите число: '))
    r = 5/x
    print(r)
except ZeroDivisionError:
    print('Вы пытаетесь делить на ноль!')
else:
    print('Всё отработало корректно') # Блок else срабатывает в том случае если при выполнении блока
#   try всё сработало корректно









